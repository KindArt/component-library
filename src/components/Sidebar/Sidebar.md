Standard:

```tsx padded
const links = [
  { title: 'Link 1', url: '/', navigate: () => {}, isActive: true },
  { title: 'Link 2', url: '/', navigate: () => {}, isActive: false },
  { title: 'Link 3', url: '/', navigate: () => {}, isActive: false },
];
<Sidebar linksTitle="Main Menu" links={links} />;
```
