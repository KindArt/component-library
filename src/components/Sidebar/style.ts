import styled from 'styled-components';
import { Colours } from '../../constants/styles';

interface ISidebarWrapperProps {
  height?: string;
  width?: string;
}

export const SidebarWrapper = styled.div<ISidebarWrapperProps>`
  display: flex;
  flex-direction: column;
  height: ${({ height }) => (height ? height : '100vh')};
  width: ${({ width }) => (width ? width : '70vw')};
`;

export const Title = styled.h1`
  margin: 30px 0 0 0;
  text-transform: uppercase;
  font-size: 0.75rem;
  color: ${Colours.TEXT_SECONDARY};
  padding-left: 20px;
`;

interface ILinkProps {
  isActive: boolean;
}

export const LinksContainer = styled.div`
  margin-top: 20px;
`;

export const Link = styled.div<ILinkProps>`
  color: ${Colours.TEXT_MAIN};
  display: flex;
  align-items: center;
  height: 45px;
  padding-left: ${({ isActive }) => (isActive ? '20px' : '24px')};

  ${({ isActive, theme }) =>
    isActive &&
    `
      border-left: 4px solid ${theme.borderColor};
      border-top: 1px solid rgba(22, 101, 216, 0.2);
      border-bottom: 1px solid rgba(22, 101, 216, 0.2);
      background-color: rgba(22, 101, 216, 0.08);
  `}

  :hover {
    cursor: pointer;
    color: ${Colours.PRIMARY};
  }
`;

Link.defaultProps = {
  theme: {
    borderColor: Colours.PRIMARY,
  },
};
