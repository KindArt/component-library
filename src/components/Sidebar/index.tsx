import React, { FC } from 'react';
import { SidebarWrapper, Title, LinksContainer, Link } from './style';

interface ILinks {
  title: string;
  url: string;
  navigate: (url: string) => void;
  order?: number;
  isActive: boolean;
}

export interface ISidebarProps {
  linksTitle?: string;
  links: ILinks[];
  height?: string;
  width?: string;
}

const renderLinks = (links: ILinks[]) => {
  const sortedLinks = links.sort((a, b) => {
    if (!a.order || !b.order) {
      return 0;
    }
    return a.order - b.order;
  });

  return sortedLinks.map(({ title, url, navigate, isActive }, index) => {
    return (
      <Link key={index} onClick={() => navigate(url)} isActive={isActive} data-testid="navigationLink">
        {title}
      </Link>
    );
  });
};

const Sidebar: FC<ISidebarProps> = ({ linksTitle, links, height, width }) => {
  return (
    <SidebarWrapper height={height} width={width}>
      <Title>{linksTitle}</Title>
      <LinksContainer>{renderLinks(links)}</LinksContainer>
    </SidebarWrapper>
  );
};

export default Sidebar;
