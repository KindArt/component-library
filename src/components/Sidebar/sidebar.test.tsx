import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import Sidebar, { ISidebarProps } from '.';

describe('<Sidebar />', () => {
  let navigateMock = jest.fn();
  const linksMock = [{ title: 'Link 1', url: '/', navigate: navigateMock, isActive: false }];

  const getSidebar = (props?: any) => {
    const defaultProps: ISidebarProps = {
      links: linksMock,
      ...props,
    };

    const utils = render(<Sidebar {...defaultProps} />);
    return { ...utils };
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render', () => {
    const { container } = getSidebar();
    expect(container).toMatchSnapshot();
  });

  describe('Visuals', () => {
    it('should have the correct amount of links', () => {
      const newProps = {
        links: [{ title: 'Link 2', url: '/', navigate: navigateMock, isActive: false }, ...linksMock],
      };

      const { queryAllByTestId } = getSidebar(newProps);
      const totalLinks = queryAllByTestId('navigationLink').length;
      expect(totalLinks).toEqual(2);
    });

    it('should have a links title if one is passed', () => {
      const newProps = { linksTitle: 'Title' };
      const { queryByText } = getSidebar(newProps);
      const title = queryByText('Title');
      expect(title).toBeTruthy();
    });
  });

  describe('Actions', () => {
    it('should call navigateMock on click', async () => {
      const { getByText } = getSidebar();
      const link = getByText('Link 1');

      await act(async () => {
        await fireEvent.click(link);
      });

      expect(navigateMock).toHaveBeenCalledTimes(1);
    });
  });
});
