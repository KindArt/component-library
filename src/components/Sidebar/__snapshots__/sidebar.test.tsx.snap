// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> should render 1`] = `
<div>
  <div
    class="sc-AxjAm mRWQf"
  >
    <h1
      class="sc-AxirZ eBheYc"
    />
    <div
      class="sc-AxiKw jrYMFl"
    >
      <div
        class="sc-AxhCb kuspGc"
        data-testid="navigationLink"
      >
        Link 1
      </div>
    </div>
  </div>
</div>
`;
